pallets/migrations/src/lib.rs

```diff
@@ -103,6 +115,88 @@ pub mod pallet {
 
 			weight
 		}
+
+		#[cfg(feature = "try-runtime")]
+		fn pre_upgrade() -> Result<(), &'static str> {
+			use frame_support::traits::OnRuntimeUpgradeHelpersExt;
+
+			let mut failed = false;
+			for migration in &T::MigrationsList::get() {
+				let migration_name = migration.friendly_name();
+				let migration_name_as_bytes = migration_name.as_bytes();
+
+				let migration_done = <MigrationState<T>>::get(migration_name_as_bytes);
+				if migration_done {
+					continue;
+				}
+				log::trace!("invoking pre_upgrade() on migration {}", migration_name);
+
+				// dump the migration name to temp storage so post_upgrade will know which
+				// migrations were performed (as opposed to skipped)
+				Self::set_temp_storage(true, migration_name);
+
+				let result = migration.pre_upgrade();
+				match result {
+					Ok(()) => {
+						log::info!("migration {} pre_upgrade() => Ok()", migration_name);
+					}
+					Err(msg) => {
+						log::error!("migration {} pre_upgrade() => Err({})", migration_name, msg);
+						failed = true;
+					}
+				}
+			}
+
+			if failed {
+				Err("One or more pre_upgrade tests failed; see output above.")
+			} else {
+				Ok(())
+			}
+		}
+
+		/// Run a standard post-runtime test. This works the same way as in a normal runtime upgrade.
+		#[cfg(feature = "try-runtime")]
+		fn post_upgrade() -> Result<(), &'static str> {
+			use frame_support::traits::OnRuntimeUpgradeHelpersExt;
+
+			// TODO: my desire to DRY all the things feels like this code is very repetitive...
+
+			let mut failed = false;
+			for migration in &T::MigrationsList::get() {
+				let migration_name = migration.friendly_name();
+
+				// we can't query MigrationState because on_runtime_upgrade() would have
+				// unconditionally set it to true, so we read a hint from temp storage which was
+				// left for us by pre_upgrade()
+				match Self::get_temp_storage::<bool>(migration_name) {
+					Some(value) => assert!(true == value, "our dummy value might as well be true"),
```

Comment:
This will panic if the assertion is false. (right?) I remember previously you argued that it's better to keep producing blocks than to panic.

Reply:
Yes, but this is in `post_upgrade()` which won't ever be included outside of testing. This code really isn't necessary; I'm just being very pedantic...

Reply:
Oh right. This is exactly where we are supposed to make assertions.